app/bundles/CampaignBundle/Tests/Functional/Executioner/InactiveExecutionerFunctionalTest.php

```diff
@@ -0,0 +1,428 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\CampaignBundle\Tests\Functional\Executioner;
+
+use Mautic\CampaignBundle\Entity\Campaign;
+use Mautic\CampaignBundle\Entity\Event;
+use Mautic\CampaignBundle\Entity\Lead as CampaignLead;
+use Mautic\CampaignBundle\Entity\LeadEventLog;
+use Mautic\CampaignBundle\Executioner\ContactFinder\Limiter\ContactLimiter;
+use Mautic\CampaignBundle\Executioner\InactiveExecutioner;
+use Mautic\CoreBundle\Test\MauticMysqlTestCase;
+use Mautic\LeadBundle\Entity\Lead;
+use Symfony\Component\Console\Output\BufferedOutput;
+
+/**
+ * Functional tests for decision event redirection scenarios.
+ * Tests redirection FROM decision events TO other event types (actions/conditions).
+ */
+class InactiveExecutionerFunctionalTest extends MauticMysqlTestCase
+{
+    public function testDecisionRedirectionToAlreadyExecutedAction(): void
+    {
+        $campaign     = $this->createCampaign();
+        $contact      = $this->createContact();
+        $campaignLead = $this->createCampaignLead($campaign, $contact);
+
+        // Create parent action event (already executed)
+        $parentEvent = $this->createActionEvent($campaign, 'Parent Event');
+
+        // Create target action that contact has ALREADY executed
+        $targetAction = $this->createActionEvent($campaign, 'Target Action');
+
+        // Create decision event (to be deleted and redirected to already executed action)
+        $originalDecision = $this->createDecisionEvent($campaign, 'Source Decision');
+        $originalDecision->setParent($parentEvent);
+
+        // Create negative child for the decision (inactive path)
+        $negativeAction = $this->createActionEvent($campaign, 'Negative Action', $originalDecision, 'no');
+        $originalDecision->addChild($negativeAction);
+
+        // Set up redirection: decision redirects to already executed action
+        $originalDecision->setDeleted(new \DateTime());
+        $originalDecision->setRedirectEvent($targetAction);
+
+        // Log: Parent event was executed successfully
+        $parentEventLog = $this->createEventLog($campaign, $parentEvent, $contact);
+
+        // Log: Target action was ALREADY executed (rotation 1)
+        $existingTargetLog = $this->createEventLog($campaign, $targetAction, $contact);
+
+        $this->em->persist($campaign);
+        $this->em->persist($contact);
+        $this->em->persist($campaignLead);
+        $this->em->persist($parentEvent);
+        $this->em->persist($targetAction);
+        $this->em->persist($originalDecision);
+        $this->em->persist($negativeAction);
+        $this->em->persist($parentEventLog);
+        $this->em->persist($existingTargetLog);
+        $this->em->flush();
+
+        // Execute decision validation (should redirect to already executed action)
+        /** @var InactiveExecutioner $inactiveExecutioner */
+        $inactiveExecutioner = self::getContainer()->get('mautic.campaign.executioner.inactive');
```

Comment:
```suggestion
        $inactiveExecutioner = self::getContainer()->get('mautic.campaign.executioner.inactive');
        \assert($inactiveExecutioner instanceof InactiveExecutioner);
```
This again repeats in this file. Move it to the setUp method.

Reply:
Done.